--- ndiswrapper-1.56/driver/usb.c	2010/09/13 08:43:15	2726
+++ ndiswrapper-1.56/driver/usb.c	2010/09/13 08:51:07	2727
@@ -191,7 +191,7 @@
 	if (wrap_urb->flags & WRAP_URB_COPY_BUFFER) {
 		USBTRACE("freeing DMA buffer for URB: %p %p",
 			 urb, urb->transfer_buffer);
-		usb_buffer_free(IRP_WRAP_DEVICE(irp)->usb.udev,
+		usb_free_coherent(IRP_WRAP_DEVICE(irp)->usb.udev,
 				urb->transfer_buffer_length,
 				urb->transfer_buffer, urb->transfer_dma);
 	}
@@ -304,7 +304,7 @@
 #endif
 		    )) {
 		urb->transfer_buffer =
-			usb_buffer_alloc(wd->usb.udev, buf_len, alloc_flags,
+			usb_alloc_coherent(wd->usb.udev, buf_len, alloc_flags,
 					 &urb->transfer_dma);
 		if (!urb->transfer_buffer) {
 			WARNING("couldn't allocate dma buf");

