--- ndiswrapper-1.56/driver/ntoskernel.h	2010/09/13 08:43:15	2726
+++ ndiswrapper-1.56/driver/ntoskernel.h	2010/09/13 08:51:07	2727
@@ -324,6 +324,12 @@
 	kmem_cache_create(name, size, align, flags, NULL)
 #endif
 
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,34)
+#define netdev_mc_count(dev) ((dev)->mc_count)
+#define usb_alloc_coherent(dev, size, mem_flags, dma) (usb_buffer_alloc((dev), (size), (mem_flags), (dma)))
+#define usb_free_coherent(dev, size, addr, dma) (usb_buffer_free((dev), (size), (addr), (dma)))
+#endif
+
 #include "winnt_types.h"
 #include "ndiswrapper.h"
 #include "pe_linker.h"

