Index: ndiswrapper-1.60/driver/ntoskernel.h
===================================================================
--- ndiswrapper-1.60.orig/driver/ntoskernel.h
+++ ndiswrapper-1.60/driver/ntoskernel.h
@@ -109,7 +109,11 @@ static cpumask_t cpumasks[NR_CPUS];
 #endif /* CONFIG_SMP */
 
 #ifndef tsk_cpus_allowed
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,3,0)
 #define tsk_cpus_allowed(tsk) (&(tsk)->cpus_allowed)
+#else
+#define tsk_cpus_allowed(tsk) ((tsk)->cpus_ptr)
+#endif
 #endif
 
 #ifndef __packed
