--- ndiswrapper-1.56/driver/winnt_types.h	2010-02-10 20:09:32.000000000 -0600
+++ ndiswrapper-1.56/driver/winnt_types.h	2011-05-26 03:36:46.496398169 -0500
@@ -138,7 +138,6 @@
 
 #endif
 
-#define packed __attribute__((packed))
 #define no_warn_unused __attribute__((unused))
 
 typedef u8	BOOLEAN;
--- ndiswrapper-1.56/driver/pe_linker.h	2010-02-10 20:09:32.000000000 -0600
+++ ndiswrapper-1.56/driver/pe_linker.h	2011-05-26 03:36:57.348262500 -0500
@@ -691,7 +691,7 @@
 
 /* Import module directory */
 
-typedef struct packed _IMAGE_IMPORT_DESCRIPTOR {
+typedef struct __packed _IMAGE_IMPORT_DESCRIPTOR {
 	union {
 		DWORD	Characteristics; /* 0 for terminating null
 					  * import descriptor */
